matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/SyncResponsePostTreatmentAggregatorHandler.kt

```diff
@@ -16,18 +16,53 @@
 
 package org.matrix.android.sdk.internal.session.sync
 
+import org.matrix.android.sdk.api.MatrixPatterns
+import org.matrix.android.sdk.internal.session.user.accountdata.DirectChatsHelper
+import org.matrix.android.sdk.internal.session.user.accountdata.UpdateUserAccountDataTask
 import javax.inject.Inject
 
 internal class SyncResponsePostTreatmentAggregatorHandler @Inject constructor(
-        private val ephemeralTemporaryStore: RoomSyncEphemeralTemporaryStore
+        private val directChatsHelper: DirectChatsHelper,
+        private val ephemeralTemporaryStore: RoomSyncEphemeralTemporaryStore,
+        private val updateUserAccountDataTask: UpdateUserAccountDataTask
 ) {
-    fun handle(synResHaResponsePostTreatmentAggregator: SyncResponsePostTreatmentAggregator) {
+    suspend fun handle(synResHaResponsePostTreatmentAggregator: SyncResponsePostTreatmentAggregator) {
         cleanupEphemeralFiles(synResHaResponsePostTreatmentAggregator.ephemeralFilesToDelete)
+        updateDirectUserIds(synResHaResponsePostTreatmentAggregator.directChatsToCheck)
     }
 
     private fun cleanupEphemeralFiles(ephemeralFilesToDelete: List<String>) {
         ephemeralFilesToDelete.forEach {
             ephemeralTemporaryStore.delete(it)
         }
     }
+
+    private suspend fun updateDirectUserIds(directUserIdsToUpdate: Map<String, String>) {
+        val directChats = directChatsHelper.getLocalDirectMessages()
+        var hasUpdate = false
+        directUserIdsToUpdate.forEach { (roomId, candidateUserId) ->
+            // consider room is a DM if referenced in the DM dictionary
+            val currentDirectUserId = directChats.firstNotNullOfOrNull { (userId, roomIds) -> userId.takeIf { roomId in roomIds } }
+            // update directUserId with the given candidateUserId if it mismatches the current one
+            if (currentDirectUserId != null && !MatrixPatterns.isUserId(currentDirectUserId)) {
+                // link roomId with the matrix id
+                directChats
+                        .getOrPut(candidateUserId) { arrayListOf() }
+                        .apply {
+                            if (!contains(roomId)) {
+                                hasUpdate = true
+                                add(roomId)
+                            }
+                        }
+
+                // remove roomId from currentDirectUserId entry
+                hasUpdate = hasUpdate or(directChats[currentDirectUserId]?.remove(roomId) == true)
```

Comment:
is it necessary to assign `hasUpdate` here?
Also I think that if `hasUpdate` is true, the part on the right of `or` will not be evaluated.

Reply:
Yes, the `remove` method will modify the direct chat dictionary if the given item has been found and removed, so I want to edit this flag in that case.

Reply:
Yes, but I was wondering if `hasUpdate` was not already equal to true in this case. It can maybe add some cleanup on some existing account data.

Reply:
Ok, I think I understand now.

Reply:
I'm not sure to understand, I use the `or` function which will always call the second part of the operation (whether hasUpdate is true or not). The flag should be updated for each modification in the direct chats dictionary. In our case, the cleanup will be performed only for members related to the given `directUserIdsToUpdate` dictionary.